# Bank NIFTY Assignment

In [60]:
# Importing Packages

import numpy as np
import pandas as pd
import datetime

import warnings
warnings.filterwarnings("ignore")

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

In [61]:
# Reading Data
data = pd.read_excel('https://internshala.com/uploads/chat-uploads/600324a823e211610818728-5191284.xlsx')
data.head()

Ticker       Date  Final Date format      Time      Open 5  \
0  BANKNIFTY_F1 2015-01-01           20150101  09:20:00  18861.3496   
1  BANKNIFTY_F1 2015-01-01           20150101  09:25:00  18808.0000   
2  BANKNIFTY_F1 2015-01-01           20150101  09:30:00  18826.9492   
3  BANKNIFTY_F1 2015-01-01           20150101  09:35:00  18825.0508   
4  BANKNIFTY_F1 2015-01-01           20150101  09:40:00  18805.0000   

       High 5       Low 5     Close 5  Volume 5  Year  
0  18867.9492  18793.0000  18808.3496     84825  2015  
1  18835.0000  18802.1992  18829.0000     40925  2015  
2  18833.4004  18820.0000  18826.0996     27000  2015  
3  18829.0996  18810.0000  18810.0000     18025  2015  
4  18808.9492  18775.0000  18803.9492     60100  2015

In [62]:
# Type of the Attributes
data.dtypes

Ticker                       object
Date                 datetime64[ns]
Final Date format             int64
Time                         object
Open 5                      float64
High 5                      float64
Low 5                       float64
Close 5                     float64
Volume 5                      int64
Year                          int64
dtype: object

In [63]:
# Number of rows and columns
data.shape

(91737, 10)

In [64]:
# Column-wise missing values
data.isnull().sum()

Ticker               0
Date                 0
Final Date format    0
Time                 0
Open 5               0
High 5               0
Low 5                0
Close 5              0
Volume 5             0
Year                 0
dtype: int64

In [65]:
# Converting the datatype of Year from int to str type
data.Year = data.Year.astype(str)
print(data.Year.dtypes)

object


In [66]:
# Converting the datatype of Date and Time to str type
data['Date'] = data['Date'].astype(str)
data['Time'] = data['Time'].astype(str)

print(data.Date.dtypes)
print(data.Time.dtypes)

object
object


In [67]:
# A new column Datetime derived from Date and Time
data['Datetime'] = data['Date'] + ' ' + data['Time']
data.Datetime.head()

0    2015-01-01 09:20:00
1    2015-01-01 09:25:00
2    2015-01-01 09:30:00
3    2015-01-01 09:35:00
4    2015-01-01 09:40:00
Name: Datetime, dtype: object

In [68]:
# Converting the datatype of Datetime from str to datetime type
data['Datetime'] = pd.to_datetime(data['Datetime'])
data.Datetime.dtypes

dtype('<M8[ns]')

In [69]:
# Parse the time. 
data.Datetime = pd.to_datetime(data.Datetime, format="%Y%m%d%H%M")
print(data.Datetime.dtypes)

datetime64[ns]


In [70]:
# Make datetime the index.
data = data.set_index('Datetime')
data.head()

Ticker        Date  Final Date format      Time  \
Datetime                                                                     
2015-01-01 09:20:00  BANKNIFTY_F1  2015-01-01           20150101  09:20:00   
2015-01-01 09:25:00  BANKNIFTY_F1  2015-01-01           20150101  09:25:00   
2015-01-01 09:30:00  BANKNIFTY_F1  2015-01-01           20150101  09:30:00   
2015-01-01 09:35:00  BANKNIFTY_F1  2015-01-01           20150101  09:35:00   
2015-01-01 09:40:00  BANKNIFTY_F1  2015-01-01           20150101  09:40:00   

                         Open 5      High 5       Low 5     Close 5  Volume 5  \
Datetime                                                                        
2015-01-01 09:20:00  18861.3496  18867.9492  18793.0000  18808.3496     84825   
2015-01-01 09:25:00  18808.0000  18835.0000  18802.1992  18829.0000     40925   
2015-01-01 09:30:00  18826.9492  18833.4004  18820.0000  18826.0996     27000   
2015-01-01 09:35:00  18825.0508  18829.0996  18810.0000  18810.0000     18025   
2015-01-01 09:40:00  18805.0000  18808.9492  18775.0000  18803.9492     60100   

                     Year  
Datetime                   
2015-01-01 09:20:00  2015  
2015-01-01 09:25:00  2015  
2015-01-01 09:30:00  2015  
2015-01-01 09:35:00  2015  
2015-01-01 09:40:00  2015

In [71]:
# Creating a new dataframe data_15_min for storing the original data grouped in 15-minute chunks. 
data_15_min = data.groupby(pd.Grouper(freq='15Min')).agg({"Open 5": "first", 
                                             "Close 5": "last", 
                                             "Low 5": "min", 
                                             "High 5": "max",
                                             "Volume 5": "sum",
                                             "Year": "first",
                                             "Date": "first",
                                             "Time": "first"})

start = datetime.time(9,15,0)
end = datetime.time(15,30,0)
data_15_min = data_15_min.between_time(start,end)

data_15_min.head()

Open 5     Close 5    Low 5      High 5  Volume 5  \
Datetime                                                                     
2015-01-01 09:15:00  18861.3496  18829.0000  18793.0  18867.9492    125750   
2015-01-01 09:30:00  18826.9492  18803.9492  18775.0  18833.4004    105125   
2015-01-01 09:45:00  18803.9492  18846.8008  18800.0  18850.0000     79575   
2015-01-01 10:00:00  18846.9004  18850.0000  18830.0  18858.5996     47300   
2015-01-01 10:15:00  18848.5508  18840.2500  18830.0  18858.0000     27800   

                     Year        Date      Time  
Datetime                                         
2015-01-01 09:15:00  2015  2015-01-01  09:20:00  
2015-01-01 09:30:00  2015  2015-01-01  09:30:00  
2015-01-01 09:45:00  2015  2015-01-01  09:45:00  
2015-01-01 10:00:00  2015  2015-01-01  10:00:00  
2015-01-01 10:15:00  2015  2015-01-01  10:15:00

In [72]:
"""Creating new columns-
TP shows Trading Price per share which is derived from adding High 5, Low 5 and Close 5.
TradedValue shows total trading price for the total volume of shares.
CumVolume shows running sum of the Volume.
CumTradedValue shows running sum of the TradedValue.
VWAP stands for Volume Weighted Average Price, derived from dividing CumTradedValue by CumVolume.
"""
def calculateVwap(data):
    data['TP'] = (data['High 5'] + data['Low 5'] + data['Close 5']) / 3.0
    data['TradedValue']  = data['TP'] * data['Volume 5']
    data['CumVolume'] = data['Volume 5'].cumsum()
    data['CumTradedValue'] = data['TradedValue'].cumsum()
    data['VWAP'] = data['CumTradedValue'] /data['CumVolume']
    return data

data_15_min = calculateVwap(data_15_min)

data_15_min = data_15_min.dropna()

data_15_min.head(30)

Open 5     Close 5       Low 5      High 5  Volume 5  \
Datetime                                                                        
2015-01-01 09:15:00  18861.3496  18829.0000  18793.0000  18867.9492    125750   
2015-01-01 09:30:00  18826.9492  18803.9492  18775.0000  18833.4004    105125   
2015-01-01 09:45:00  18803.9492  18846.8008  18800.0000  18850.0000     79575   
2015-01-01 10:00:00  18846.9004  18850.0000  18830.0000  18858.5996     47300   
2015-01-01 10:15:00  18848.5508  18840.2500  18830.0000  18858.0000     27800   
2015-01-01 10:30:00  18843.7500  18840.0000  18825.0000  18843.7500     14150   
2015-01-01 10:45:00  18836.0000  18845.0000  18828.0000  18846.2500     13325   
2015-01-01 11:00:00  18846.0000  18840.0996  18832.0000  18848.6992     20075   
2015-01-01 11:15:00  18842.0000  18870.0000  18837.4004  18889.9004     86025   
2015-01-01 11:30:00  18870.0000  18855.0508  18851.9492  18878.0000     20875   
2015-01-01 11:45:00  18856.0000  18867.0000  18855.0508  18877.6992     24025   
2015-01-01 12:00:00  18867.0000  18870.5000  18865.0000  18889.0000     40600   
2015-01-01 12:15:00  18870.5508  18866.0000  18862.0508  18880.0000     14475   
2015-01-01 12:30:00  18862.5000  18864.9004  18850.0000  18867.9004     20600   
2015-01-01 12:45:00  18862.0000  18871.0000  18855.0508  18873.0000      8500   
2015-01-01 13:00:00  18872.0000  18864.0000  18857.4492  18872.0000      8075   
2015-01-01 13:15:00  18865.0000  18861.9004  18857.1504  18869.7500      7400   
2015-01-01 13:30:00  18860.4492  18870.0000  18853.0000  18879.5996     17300   
2015-01-01 13:45:00  18869.0000  18880.0000  18861.1504  18880.8008     14025   
2015-01-01 14:00:00  18880.0000  18893.0000  18870.0000  18909.9492     74300   
2015-01-01 14:15:00  18892.9004  18884.0000  18875.1992  18893.0000     21200   
2015-01-01 14:30:00  18884.0000  18902.9004  18876.0508  18920.0000     82600   
2015-01-01 14:45:00  18900.3496  18913.0000  18890.0508  18929.9492     96100   
2015-01-01 15:00:00  18914.0000  18903.0000  18892.5996  18914.0000     43250   
2015-01-01 15:15:00  18903.0000  18900.0000  18893.5996  18905.9004     62575   
2015-01-02 09:15:00  18899.8008  18996.0000  18898.0000  18996.5508    220825   
2015-01-02 09:30:00  18996.5508  19050.0000  18979.0996  19050.0000    222125   
2015-01-02 09:45:00  19051.8496  19119.0000  19046.0000  19130.0000    290550   
2015-01-02 10:00:00  19118.0000  19174.0000  19106.2500  19180.0000    195650   
2015-01-02 10:15:00  19170.5000  19175.0996  19154.5000  19188.0000    119000   

                     Year        Date      Time            TP   TradedValue  \
Datetime                                                                      
2015-01-01 09:15:00  2015  2015-01-01  09:20:00  18829.983067  2.367870e+09   
2015-01-01 09:30:00  2015  2015-01-01  09:30:00  18804.116533  1.976783e+09   
2015-01-01 09:45:00  2015  2015-01-01  09:45:00  18832.266933  1.498578e+09   
2015-01-01 10:00:00  2015  2015-01-01  10:00:00  18846.199867  8.914253e+08   
2015-01-01 10:15:00  2015  2015-01-01  10:15:00  18842.750000  5.238284e+08   
2015-01-01 10:30:00  2015  2015-01-01  10:30:00  18836.250000  2.665329e+08   
2015-01-01 10:45:00  2015  2015-01-01  10:45:00  18839.750000  2.510397e+08   
2015-01-01 11:00:00  2015  2015-01-01  11:00:00  18840.266267  3.782183e+08   
2015-01-01 11:15:00  2015  2015-01-01  11:15:00  18865.766933  1.622928e+09   
2015-01-01 11:30:00  2015  2015-01-01  11:30:00  18861.666667  3.937373e+08   
2015-01-01 11:45:00  2015  2015-01-01  11:45:00  18866.583333  4.532697e+08   
2015-01-01 12:00:00  2015  2015-01-01  12:00:00  18874.833333  7.663182e+08   
2015-01-01 12:15:00  2015  2015-01-01  12:15:00  18869.350267  2.731338e+08   
2015-01-01 12:30:00  2015  2015-01-01  12:30:00  18860.933600  3.885352e+08   
2015-01-01 12:45:00  2015  2015-01-01  12:45:00  18866.350267  1.603640e+08   
2015-01-01 13:00:00  2015  2015-01-01  13:00:00  18864.483067  1.523307e+08   
2015-01-01

In [73]:
# Number of rows and columns in data_15_min
data_15_min.shape

(30640, 13)

In [74]:
# Column-wise missing values
data_15_min.isnull().sum()

Open 5            0
Close 5           0
Low 5             0
High 5            0
Volume 5          0
Year              0
Date              0
Time              0
TP                0
TradedValue       0
CumVolume         0
CumTradedValue    0
VWAP              0
dtype: int64

In [75]:
# Here, we'll create a dataframe for storing the Segregated profit and loss Date-wise.

# Making a list of unique Dates.
dates = data_15_min['Date'].unique().tolist()

# Creating an empty array to store the results.
profit_loss_arr = []

# Iterating over Dates.
for d in dates:
#   Creating a variable datewise_data to store the data for each time duration.
    datewise_data = data_15_min[d]
#   Creating some variables to store the values of total volume, total trading price and stop loss respectively.
    total_vol = 0
    total_trading_price = 0
    stop_loss = 0
    
    for t in range(1, len(datewise_data)-1):
#       Here is some conditions to know the trade of shares.
#       First, if candle closes above VWAP, we will buy.
        if datewise_data['Close 5'][t] > datewise_data['VWAP'][t]:
#           Total volume increases and trading price decreases while buying.
            total_vol += datewise_data['Volume 5'][t]
            total_trading_price -= datewise_data['Volume 5'][t] * datewise_data['TP'][t]
            stop_loss = datewise_data['Low 5'][t-1]
        
#       Second, if candle closes below VWAP, we will sell.
        elif datewise_data['Close 5'][t] < datewise_data['VWAP'][t] or datewise_data['Close 5'][t] < stop_loss:
#           Total volume decreases and trading price increases while selling.
            total_vol -= datewise_data['Volume 5'][t]
            total_trading_price += datewise_data['Volume 5'][t] * datewise_data['TP'][t]

#   Atlast, we'll sell all the remaining shares because Intraday Trades are to be taken and mandatorily closed in the same day.
    if total_vol > 0:
        total_trading_price += total_vol * datewise_data['TP'][-1]
    else:
        total_trading_price -= total_vol * datewise_data['TP'][-1]
#   Appending all the results to the array.
    profit_loss_arr.append([datewise_data['Year'][-1], datewise_data['Date'][-1], total_trading_price])

# Creating a new dataframe to store the resulted profit and loss.
Profit_Loss_df = pd.DataFrame(data = profit_loss_arr, columns = ['Year', 'Date', 'ProfitLoss'])
    
Profit_Loss_df

Year        Date    ProfitLoss
0     2015  2015-01-01  1.067065e+07
1     2015  2015-01-02  1.093157e+08
2     2015  2015-01-05 -4.286471e+07
3     2015  2015-01-06  1.369261e+11
4     2015  2015-01-07  1.158540e+11
5     2015  2015-01-08  7.221796e+10
6     2015  2015-01-09  6.768563e+10
7     2015  2015-01-12  3.016762e+07
8     2015  2015-01-13 -2.740313e+08
9     2015  2015-01-14  9.112544e+10
10    2015  2015-01-15 -6.496859e+07
11    2015  2015-01-16  1.071895e+07
12    2015  2015-01-19  2.931437e+07
13    2015  2015-01-20  3.598985e+08
14    2015  2015-01-21 -9.765628e+07
15    2015  2015-01-22  1.070899e+08
16    2015  2015-01-23  2.822893e+06
17    2015  2015-01-27  7.196906e+08
18    2015  2015-01-28 -6.249526e+08
19    2015  2015-01-29  2.122200e+08
20    2015  2015-01-30 -9.253528e+08
21    2015  2015-02-02  2.677754e+08
22    2015  2015-02-03 -1.192784e+09
23    2015  2015-02-04  9.089399e+10
24    2015  2015-02-05  8.534198e+10
25    2015  2015-02-06  1.262252e+11
26    2015  2015-02-09  8.763563e+10
27    2015  2015-02-10  1.772685e+11
28    2015  2015-02-11  8.115816e+10
29    2015  2015-02-12  1.223067e+11
30    2015  2015-02-13 -9.852263e+07
31    2015  2015-02-16 -3.335715e+08
32    2015  2015-02-18 -9.020959e+07
33    2015  2015-02-19  1.150022e+11
34    2015  2015-02-20  1.461498e+11
35    2015  2015-02-23  8.694803e+10
36    2015  2015-02-24  1.261752e+11
37    2015  2015-02-25  1.129539e+11
38    2015  2015-02-26  1.085453e+11
39    2015  2015-02-27  1.212602e+11
40    2015  2015-02-28  2.043994e+09
41    2015  2015-03-02  3.735648e+08
42    2015  2015-03-03  1.049103e+08
43    2015  2015-03-04 -2.100807e+09
44    2015  2015-03-05  7.382824e+08
45    2015  2015-03-09 -2.438714e+08
46    2015  2015-03-10  1.173088e+11
47    2015  2015-03-11  1.227283e+11
48    2015  2015-03-12  6.418308e+10
49    2015  2015-03-13  1.321837e+11
50    2015  2015-03-16  1.041635e+11
51    2015  2015-03-17  1.318271e+11
52    2015  2015-03-18  1.189663e+11
53    2015  2015-03-19  5.622591e+10
54    2015  2015-03-20  1.178708e+11
55    2015  2015-03-23  9.690025e+10
56    2015  2015-03-24  1.496368e+11
57    2015  2015-03-25  8.204970e+10
58    2015  2015-03-26  1.327618e+11
59    2015  2015-03-27  1.267504e+11
60    2015  2015-03-30  8.312721e+10
61    2015  2015-03-31  9.737916e+10
62    2015  2015-04-01  1.061373e+11
63    2015  2015-04-06  8.980453e+10
64    2015  2015-04-07  1.376193e+11
65    2015  2015-04-08  8.835429e+10
66    2015  2015-04-09  1.815936e+11
67    2015  2015-04-10  7.452978e+10
68    2015  2015-04-13  7.003175e+10
69    2015  2015-04-15  9.573984e+10
70    2015  2015-04-16  1.107797e+11
71    2015  2015-04-17  8.835783e+10
72    2015  2015-04-20  1.256524e+11
73    2015  2015-04-21  1.112921e+11
74    2015  2015-04-22  1.519499e+11
75    2015  2015-04-23  1.353199e+11
76    2015  2015-04-24  9.493127e+10
77    2015  2015-04-27  9.700194e+10
78    2015  2015-04-28  1.317859e+11
79    2015  2015-04-29  1.039776e+11
80    2015  2015-04-30  1.051904e+11
81    2015  2015-05-04  7.425166e+10
82    2015  2015-05-05  7.645423e+10
83    2015  2015-05-06  1.520311e+11
84    2015  2015-05-07  1.379127e+11
85    2015  2015-05-08  1.093637e+11
86    2015  2015-05-11  9.487080e+10
87    2015  2015-05-12  1.012230e+11
88    2015  2015-05-13  1.595179e+11
89    2015  2015-05-14  1.043575e+11
90    2015  2015-05-15  6.693271e+10
91    2015  2015-05-18  7.286354e+10
92    2015  2015-05-19  1.043499e+11
93    2015  2015-05-20  6.860885e+10
94    2015  2015-05-21  8.248627e+10
95    2015  2015-05-22  1.180946e+11
96    2015  2015-05-25  6.441192e+10
97    2015  2015-05-26  6.235382e+10
98    2015  2015-05-27  1.008702e+11
99    2015  2015-05-28  9.529921e+10
100   2015  2015-05-29  8.025916e+10
101   2015  2015-06-01  6.958636e+10
102   2015  2015-06-02  1.680372e+11
103   2015  2015-06-03  1.150600e+11
104   2015  2015-06-04  1.214708e+11
105   2015  2015-06-05  9.075127e+10
106   2015  2015-06-08  7.404831e+10
107   2015

In [76]:
# Computing the yearwise profit/loss and storing it in a dataframe named yearwise_profit_loss.
yearwise_profit_loss = Profit_Loss_df.groupby(['Year'])
yearwise_profit_loss = pd.DataFrame(yearwise_profit_loss['ProfitLoss'].sum())
yearwise_profit_loss = yearwise_profit_loss.reset_index()
yearwise_profit_loss


Year    ProfitLoss
0  2015  1.968833e+13
1  2016  9.296279e+12
2  2017  6.908574e+09
3  2018  7.431940e+09
4  2019  2.247823e+08

In [77]:
# Computing the maximum profit making year
best_performing_year = max(yearwise_profit_loss['ProfitLoss'])
best_performing_year

# It is clearly seen that 

19688328345799.0

In [78]:
# Overall Profit/Loss
overall_profit_loss = yearwise_profit_loss['ProfitLoss'].sum()
overall_profit_loss

28999172219312.96

- Profit making year is 2015
- The Bank NIFTY is making profit of 28999172219312.96 over all the years. 